### PR TITLE
Support differentiable coefficients for observables

### PR DESCRIPTION
**Pull request by albi3ro**
**Context:**

With legacy operator arithmetic, we supported trainable coefficients using the `hadamard_grad` transform. This doesn't really work anymore with generic operator arithmetic.  

**Description of the Change:**

Uses the `split_to_single_terms` transform in gradient preprocessing if any of the observables have trainable coefficients.

In order to get this to pass tests, various bugs in `split_non_commuting` needed fixing.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

[sc-71490]

### CODE REVIEW

File: pennylane/workflow/execution.py [hunk 1]
```diff
@@ -483,7 +483,7 @@ def cost_fn(params, x):
 
     is_gradient_transform = isinstance(diff_method, qml.transforms.core.TransformDispatcher)
```
Comment (albi3ro) [inline pennylane/workflow/execution.py:484]:
```suggestion
```

File: tests/interfaces/test_jax.py [hunk 2]
```diff
@@ -820,6 +819,8 @@ def test_multiple_hamiltonians_trainable(self, execute_kwargs, cost_fn, shots):
         """Test hamiltonian with trainable parameters."""
         if execute_kwargs["diff_method"] == "adjoint":
             pytest.skip("trainable hamiltonians not supported with adjoint")
+        if shots.has_partitioned_shots:
+            pytest.xfail("there are problems here.")
```
Comment (albi3ro) [inline tests/interfaces/test_jax.py:823]:
```suggestion
```

File: pennylane/gradients/vjp.py [hunk 3]
```diff
@@ -129,6 +129,8 @@ def compute_vjp_single(dy, jac, num=None):
         if num == 1:
             jac = qml.math.squeeze(jac)
         jac = qml.math.reshape(jac, (-1, 1))
+        print("dy_row: ", dy_row)
+        print("jac: ", jac)
```
Comment (albi3ro) [inline pennylane/gradients/vjp.py:133]:
```suggestion
```

File: pennylane/transforms/split_non_commuting.py [hunk 4]
```diff
@@ -572,8 +589,7 @@ def _processing_fn_no_grouping(
     res: ResultBatch,
     single_term_obs_mps: dict[MeasurementProcess, tuple[list[int], list[Union[float, TensorLike]]]],
     offsets: list[Union[float, TensorLike]],
-    shots: Shots,
-    batch_size: int,
+    batch_size: None | int,
```
Comment (albi3ro) [inline pennylane/transforms/split_non_commuting.py:592]:
```suggestion
    batch_size: Union[None, int],
```

File: pennylane/ops/op_math/sprod.py [hunk 5]
```diff
@@ -148,7 +147,7 @@ def __init__(
         elif (base_pauli_rep := getattr(self.base, "pauli_rep", None)) and (
             self.batch_size is None
         ):
-            scalar = copy(self.scalar)
+            # scalar = copy(self.scalar)
```
Comment (albi3ro) [inline pennylane/ops/op_math/sprod.py:150]:
```suggestion
```

File: tests/gradients/parameter_shift/test_parameter_shift.py [hunk 6]
```diff
@@ -3462,33 +3461,21 @@ def test_trainable_coeffs(self, mocker, tol, broadcast):
         tapes, fn = qml.gradients.param_shift(tape, broadcast=broadcast)
         # two (broadcasted if broadcast=True) shifts per rotation gate
         # one circuit per trainable H term
-        assert len(tapes) == (2 + 2 if broadcast else 2 * 2 + 2)
-        assert [t.batch_size for t in tapes] == ([2, 2, None, None] if broadcast else [None] * 6)
-        spy.assert_called()
+        assert len(tapes) == (2 if broadcast else 2 * 2)
```
Comment (astralcai) [inline tests/gradients/parameter_shift/test_parameter_shift.py:3464]: Why has the expected number of tapes changed here?

Comment (albi3ro) [reply tests/gradients/parameter_shift/test_parameter_shift.py:3464]: We used to go an extra execution per term using `hamiltonian_grad`.  So it we wanted to get the derivative of `expval(2*X(0))`, we would do a bonus measurement of `expval(X(0))`.

**Comment by astralcai** [comment]
> With legacy operator arithmetic, we supported trainable coefficients using the hadamard_grad transform. This doesn't really work anymore with generic operator arithmetic.

Should we deprecate/remove `hamiltonian_grad`?

**Comment by albi3ro** [comment]
> > With legacy operator arithmetic, we supported trainable coefficients using the hadamard_grad transform. This doesn't really work anymore with generic operator arithmetic.
> 
> Should we deprecate/remove `hamiltonian_grad`?

Potentially worth considering now.

File: tests/workflow/interfaces/execute/test_torch.py [hunk 7]
```diff
@@ -871,9 +877,10 @@ def test_multiple_hamiltonians_trainable(self, execute_kwargs, cost_fn, shots):
 
         res = torch.hstack(torch.autograd.functional.jacobian(cost_fn, (weights, coeffs1, coeffs2)))
         expected = self.cost_fn_jacobian(weights, coeffs1, coeffs2)
+        print(res)
+        print(expected)
```
Comment (astralcai) [inline tests/workflow/interfaces/execute/test_torch.py:881]:
```suggestion
```

File: tests/workflow/interfaces/execute/test_torch.py [hunk 8]
```diff
@@ -854,15 +860,15 @@ def test_multiple_hamiltonians_trainable(self, execute_kwargs, cost_fn, shots):
...
-            pytest.xfail(reason="parameter shift derivatives do not yet support sums.")
 
         coeffs1 = torch.tensor([0.1, 0.2, 0.3], requires_grad=True)
         coeffs2 = torch.tensor([0.7], requires_grad=True)
         weights = torch.tensor([0.4, 0.5], requires_grad=True)
 
         res = cost_fn(weights, coeffs1, coeffs2)
         expected = self.cost_fn_expected(weights, coeffs1, coeffs2)
+        print(res)
+        print(expected)
```
Comment (astralcai) [inline tests/workflow/interfaces/execute/test_torch.py:871]:
```suggestion
```

File: tests/workflow/interfaces/execute/test_jax.py [hunk 9]
```diff
@@ -280,6 +280,7 @@ def cost(params):
                 [qml.RY(jnp.array(0.5), 0)], [qml.probs(wires=[0, 1])], shots=shots
             )
             res = execute([tape1, tape2, tape3, tape4], device, **execute_kwargs)
+            print(res)
```
Comment (astralcai) [inline tests/workflow/interfaces/execute/test_jax.py:283]:
```suggestion
```

File: tests/gradients/parameter_shift/test_parameter_shift_shot_vec.py [hunk 10]
```diff
@@ -2078,25 +2078,25 @@ def test_no_trainable_coeffs(self, mocker, broadcast, tol):
             b * np.cos(x) * np.cos(y) - c * np.cos(y) * np.sin(x),
         ]
         for res in all_res:
-            assert isinstance(res, tuple)
+            print(res)
+            # assert isinstance(res, tuple)
```
Comment (astralcai) [inline tests/gradients/parameter_shift/test_parameter_shift_shot_vec.py:2082]:
```suggestion
```

File: doc/releases/changelog-dev.md [hunk 11]
```diff
@@ -6,6 +6,9 @@
 
 <h3>Improvements 🛠</h3>
 
+* The coefficients of observables now has improved differentiation.
```
Comment (andrijapau) [inline doc/releases/changelog-dev.md:9]: Perhaps?
```suggestion
* The coefficients of observables now have improved differentiability.
```

File: pennylane/devices/reference_qubit.py [hunk 12]
```diff
@@ -70,6 +70,7 @@ def simulate(tape: qml.tape.QuantumTape, seed=None) -> qml.typing.Result:
             results.append(
                 tuple(mp.process_samples(sub_samples, tape.wires) for mp in tape.measurements)
             )
+
```
Comment (andrijapau) [inline pennylane/devices/reference_qubit.py:73]:
```suggestion
```

**Review by astralcai — APPROVED** [review]

**Review by andrijapau — APPROVED** [review]
